common/src/main/java/io/netty/util/internal/DefaultPriorityQueue.java

```diff
@@ -0,0 +1,291 @@
+        T moved = queue[i] = queue[size];
+        queue[size] = null;
+        // priorityQueueIndex will be updated below in bubbleUp or bubbleDown
+
+        // Make sure the moved node still preserves the min-heap properties.
+        if (comparator.compare(node, moved) < 0) {
+            bubbleDown(i, moved);
+        } else {
+            bubbleUp(i, moved);
+        }
+        return true;
+    }
+
+    @Override
+    public void priorityChanged(T node) {
+        int i = node.priorityQueueIndex(this);
+        if (!contains(node, i)) {
+            return;
+        }
+
+        // Preserve the min-heap property by comparing the new priority with parents/children in the heap.
+        if (i == 0) {
+            bubbleDown(i, node);
+        } else {
+            // Get the parent to see if min-heap properties are violated.
+            int iParent = (i - 1) >>> 1;
+            T parent = queue[iParent];
+            if (comparator.compare(node, parent) < 0) {
+                bubbleUp(i, node);
+            } else {
+                bubbleDown(i, node);
+            }
+        }
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(queue, size);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <X> X[] toArray(X[] a) {
+        if (a.length < size) {
+            return (X[]) Arrays.copyOf(queue, size, a.getClass());
+        }
+        System.arraycopy(queue, 0, a, 0, size);
+        if (a.length > size) {
+            a[size] = null;
+        }
+        return a;
+    }
+
+    /**
+     * This iterator does not return elements in any particular order.
+     */
+    @Override
+    public Iterator<T> iterator() {
+        return new PriorityQueueIterator();
+    }
+
+    private final class PriorityQueueIterator implements Iterator<T> {
+        private int index;
+
+        @Override
+        public boolean hasNext() {
+            return index < size;
+        }
+
+        @Override
+        public T next() {
+            if (index >= size) {
+                throw new NoSuchElementException();
+            }
+
+            return queue[index++];
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException("remove");
+        }
+    }
+
+    private boolean contains(PriorityQueueNode node, int i) {
+        return i >= 0 && i < size && node.equals(queue[i]);
+    }
+
+    private void bubbleDown(int k, T node) {
+        final int half = size >>> 1;
+        while (k < half) {
+            // Compare node to the children of index k.
+            int iChild = (k << 1) + 1;
+            T child = queue[iChild];
+
+            // Make sure we get the smallest child to compare against.
+            int rightChild = iChild + 1;
+            if (rightChild < size && comparator.compare(child, queue[rightChild]) > 0) {
+                child = queue[iChild = rightChild];
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract the assignment out of this expression. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)